vignettes/potential-optimizers.Rmd

```diff
@@ -0,0 +1,670 @@
+
+#### Unoptimized Code
+
+```{r pre_allocation_code_og}
+pre_mem <- function(n) {
+  vec <- NULL
+  for (i in seq_len(n)) {
+    vec[i] <- i
+  }
+}
+```
+
+#### Proposed Optimized Code
+
+```{r pre_allocation_code_op}
+pre_mem_opt <- function(n) {
+  vec <- vector(mode = "numeric", length = n)
+  for (i in seq_len(n)) {
+    vec[i] <- i
+  }
+}
+
+pre_mem_opt_unknown_type <- function(n) {
+  vec <- vector(length = n)
+  for (i in seq_len(n)) {
+    vec[i] <- i
+  }  
+}
+
+```
+
+### Benchmark
+
+```{r pre_allocate_benchmark, echo = TRUE, warning = FALSE, message= FALSE}
+library("ggplot2")
+library("microbenchmark")
+n <- 100000
+autoplot(microbenchmark(pre_mem(n), pre_mem_opt(n), pre_mem_opt_unknown_type(n)))
+```
+
+### Drawbacks
+
+  **To-Do**
+
+## 4. Vectorization
+
+### Idea
+A golden rule in R programming is to access the underlying C/Fortran routines as quickly as possible; the fewer functions calls required to achieve this, the better.Many R functions are therefore vectorised, that is the function’s inputs and/or outputs naturally work with vectors, reducing the number of function calls required.
+
+### Code Examples
+
+#### Unoptimized Code
+
+```{r vectorization_code_og}
+non_vectorized <- function(n) {
+  v1 <- seq_len(n)
+  v2 <- length(seq.int(n+2, n+1000, 2))
+  res <- vector(length = length(v1))
+  for (i in seq_len(n)) {
+    res[i] <- v1[i] + v2[i]
+  }
+}
+```
+
+#### Proposed Optimized Code
+
+```{r vectorization_code_op}
+vectorized <- function(n) {
+  v1 <- seq_len(n)
+  v2 <- length(seq.int(n+2, n+1000, 2))
+  res <- v1 + v2
+}
+
+```
+
+### Benchmark
+
+```{r vectorized_benchmark, echo = TRUE, warning = FALSE, message= FALSE}
+library("ggplot2")
+library("microbenchmark")
+n <- 10000
+autoplot(microbenchmark(non_vectorized(n), vectorized(n)))
+```
+
+### Drawbacks
+
+  **To-Do**
+
+## 5. Caching
+
+### Idea
+The idea of caching is similar to the concept of dynamic programming where the results of repeating sub-problems of a recursion are stored/cached and simply called every time instead of recomputating.
+
+### Code Examples
+
+#### Unoptimized Code
+
+```{r cache_code_og, echo=TRUE, warning=FALSE}
+
+library("memoise")
+
+fibR <- function(x) {
+  if (x == 0) return(0)
+  if (x == 1) return(1)
+  Recall(x - 1) + Recall(x - 2)
+}
+```
+
+#### Proposed Optimized Code
+
+```{r cache_code_op, echo=TRUE, warning=FALSE}
+mem_fibR <- memoise::memoise(fibR)
+```
+
+### Benchmark
+
+```{r cache_benchmark, warning=FALSE, message=FALSE}
+library("ggplot2")
+library("microbenchmark")
+n <- 20
+autoplot(microbenchmark(fibR(n), mem_fibR(n)))
+```
+
+### Drawbacks
+
+  **To-Do**
+
+## 6. Reversing elements
```

Comment:
It is too specific. It is optimizing an algorithm, not an optimizer technique. We should remove it for now.